# problem statement 

### 1.IMPORT REQUIRED LIBRARYS



In [220]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os 
from sklearn.model_selection import train_test_split

In [128]:
 pd.set_option('display.float_format', '{:.2f}'.format)

###  2.READING THE DATA 

In [5]:
def get_data():

    #Noida  city csv file:
    jaipur=pd.read_csv('jaipur.csv')
    
    #Noida  city csv file:
    bhopal=pd.read_csv('bhopal.csv')

    
    #Noida  city csv file:
    vadodara=pd.read_csv('vadodara.csv')

    
    #Noida  city csv file:
    surat=pd.read_csv('surat.csv')

    
    #Noida  city csv file:
    kanpur=pd.read_csv('kanpur.csv')

    
    #Noida  city csv file:
    Guwahati=pd.read_csv('Guwahati.csv')

     #Noida  city csv file:
    Faridabad=pd.read_csv('Faridabad.csv')

    


    

    #Noida  city csv file:
    Lucknow=pd.read_csv('Lucknow.csv')

    #indore  city csv file:
    indore=pd.read_csv('indore.csv')

    #Noida  city csv file:
    Noida=pd.read_csv('Noida.csv')

    #Ghaziyabad  city csv file:
    Ghaziyabad=pd.read_csv('Ghaziyabad.csv')

    # Ahmadabad city csv file:
    Ahmadabad=pd.read_csv('Ahmadabad.csv')
    

    # kolkata city csv file:
    kolkata=pd.read_csv('kolkata.csv')

    
    # Hydrabadcity csv file:
    hyderabad=pd.read_csv('hyderabad.csv')
    
    
    # gurgaoncity csv file:
    gurgaon=pd.read_csv('gurgaon.csv')

    # Delhi city csv file:
    Delhi=pd.read_csv('Delhi.csv')
    
    # Chennai city csv file:
    Chennai=pd.read_csv('Chennai.csv')

    # Banglore city csv file:
    Banglore=pd.read_csv('Banglore.csv')
    
    # pune city csv file:
    pune=pd.read_csv('pune.csv')
    
    # mumbai city csv file:
    mumbai=pd.read_csv('mumbai.csv')

   ### merge alll above city into one dataframe:
    return  pd.concat([mumbai,pune,Banglore,Chennai,Delhi,gurgaon,hyderabad,kolkata,Ahmadabad,Ghaziyabad,Noida,indore,Lucknow,jaipur,bhopal,vadodara,surat,kanpur,Guwahati,Faridabad],ignore_index=True)

    
    

In [6]:
data=get_data()

In [7]:
data

broker_consultancy    Rent  Builtup Area  appartment_type  \
0      Kasturi Developers\nBUILDER\n-  20,000          1132                2   
1      Kasturi Developers\nBUILDER\n-  20,000          1200                2   
2      Kasturi Developers\nBUILDER\n-  30,000          1700                3   
3              Seller\nVERIFIED OWNER  28,500           750                2   
4              Seller\nVERIFIED OWNER  50,000          1400                3   
...                               ...     ...           ...              ...   
82011          Seller\nVERIFIED OWNER  11,000          1100                2   
82012          Seller\nVERIFIED OWNER   9,000           900                2   
82013          Seller\nVERIFIED OWNER  22,000          2727                4   
82014          Seller\nVERIFIED OWNER  11,000           600                2   
82015          Seller\nVERIFIED OWNER  40,000          1900                3   

              Deposit    Bathrooms            Facing                 Address  \
0          No Deposit  2 bathrooms  NorthEast facing            Ulwe, Mumbai   
1          No Deposit  2 bathrooms  NorthEast facing            Ulwe, Mumbai   
2          No Deposit  3 bathrooms  NorthEast facing            Ulwe, Mumbai   
3          No Deposit  2 bathrooms               NaN  Kandivali West, Mumbai   
4          No Deposit  3 bathrooms       East facing         Chembur, Mumbai   
...               ...          ...               ...                     ...   
82011  11,000 deposit  2 bathrooms               NaN     Sector 2, Faridabad   
82012   9,000 deposit  2 bathrooms               NaN    Sector 91, Faridabad   
82013  22,000 deposit  4 bathrooms               NaN    Sector 76, Faridabad   
82014      No Deposit  2 bathrooms               NaN    Sector 88, Faridabad   
82015      No Deposit  3 bathrooms  NorthEast facing    Sector 79, Faridabad   

              Apartment Name          Status  
0                  EV Castle     Unfurnished  
1                        NaN     Unfurnished  
2                        NaN  Semi-Furnished  
3             Atul Blue Arch     Unfurnished  
4               Godrej Prime  Semi-Furnished  
...                      ...             ...  
82011                    NaN     Unfurnished  
82012    Gupta Surya Nagar 2  Semi-Furnished  
82013                    NaN  Semi-Furnished  
82014                    NaN  Semi-Furnished  
82015  Omaxe Royal Residency  Semi-Furnished  

[82016 rows x 10 columns]

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 82016 entries, 0 to 82015
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   broker_consultancy  82015 non-null  object
 1   Rent                82016 non-null  object
 2   Builtup Area        82016 non-null  int64 
 3   appartment_type     82016 non-null  int64 
 4   Deposit             82016 non-null  object
 5   Bathrooms           81672 non-null  object
 6   Facing              28451 non-null  object
 7   Address             82016 non-null  object
 8   Apartment Name      32734 non-null  object
 9   Status              82016 non-null  object
dtypes: int64(2), object(8)
memory usage: 6.3+ MB


-   The data set contains 82,016 rows and 10 features
-   column "Bathrooms ,Facing,Appartment_Name," have lots of  missing values
-   The data type of some features are not appropriate

### 3.PREMLINARY DATA ANALYSIS

#### 3.1Check for datatype


In [9]:
data.sample(150)

broker_consultancy    Rent  Builtup Area  appartment_type  \
69878          Seller\nVERIFIED OWNER  14,000           445                1   
8749      Richie Properties\nAGENT\n-  30,000           952                2   
43831         Modern Spaces\nAGENT\n-  85,000          2000                3   
13839       Vijay Umbratkar\nAGENT\n-  32,000           545                1   
51623          Seller\nVERIFIED OWNER  19,000           450                1   
...                               ...     ...           ...              ...   
28622          Seller\nVERIFIED OWNER  18,000          1000                2   
24975          Seller\nVERIFIED OWNER   6,500           350                1   
43382          Seller\nVERIFIED OWNER  39,000          1100                3   
14516  Airoli Flat Services\nAGENT\n-   6,000           400                1   
54485          Seller\nVERIFIED OWNER  16,000          1020                1   

              Deposit    Bathrooms       Facing                     Address  \
69878      No Deposit  1 bathrooms          NaN          Ghodsar, Ahmedabad   
8749       No Deposit  2 bathrooms  West facing          Thane West, Mumbai   
43831      No Deposit  4 bathrooms          NaN          Vasant Kunj, Delhi   
13839     1 L deposit  1 bathrooms          NaN              Wadala, Mumbai   
51623      No Deposit  1 bathrooms  West facing         Sector 36A, Gurgaon   
...               ...          ...          ...                         ...   
28622      No Deposit  2 bathrooms          NaN  byatarayanapura, Bangalore   
24975      No Deposit  1 bathrooms          NaN      J. P. Nagar, Bangalore   
43382      No Deposit  2 bathrooms          NaN            Janakpuri, Delhi   
14516  20,000 deposit  1 bathrooms          NaN              Airoli, Mumbai   
54485      No Deposit          NaN          NaN        PALAM VIHAR, Gurgaon   

                                     Apartment Name          Status  
69878                                           NaN  Semi-Furnished  
8749   Lodha Splendora Platino E To G Vivant A To D       Furnished  
43831                         DDA Flats Vasant Kunj     Unfurnished  
13839                            Ajmera Bhakti Park  Semi-Furnished  
51623                                        AVL 36       Furnished  
...                                             ...             ...  
28622                                           NaN  Semi-Furnished  
24975                                           NaN  Semi-Furnished  
43382                                           NaN  Semi-Furnished  
14516                                  Ekvira Topaz     Unfurnished  
54485                                           NaN  Semi-Furnished  

[150 rows x 10 columns]

In [10]:
data.dtypes

broker_consultancy    object
Rent                  object
Builtup Area           int64
appartment_type        int64
Deposit               object
Bathrooms             object
Facing                object
Address               object
Apartment Name        object
Status                object
dtype: object

#### 3.2Check For Duplicates

In [11]:
data.duplicated().sum()

68994

In [12]:
duplicated_rows=(
    data.loc[data.duplicated(keep=False)].sort_values(["Rent","Builtup Area","appartment_type","Bathrooms","Facing","Address","Apartment Name","Status"])
)

In [13]:
duplicated_rows.sample(150)

broker_consultancy    Rent  Builtup Area  appartment_type  \
33957         Seller\nVERIFIED OWNER  48,000          1020                3   
45489        Gaurav Dahiya\nAGENT\n-  28,000          1100                3   
55208  SYNERGY CONSULTANTS\nAGENT\n-  42,000          1886                3   
10108              Azuroin\nAGENT\n-   2.5 L          1965                3   
66484         Mondal Homes\nAGENT\n0  40,000          2200                4   
...                              ...     ...           ...              ...   
67496         Seller\nVERIFIED OWNER  27,000          1200                3   
26662         Seller\nVERIFIED OWNER  18,000          1000                2   
30800         Seller\nVERIFIED OWNER  18,000           550                1   
29682         Seller\nVERIFIED OWNER  18,000          1000                2   
46055    Adarsh Associates\nAGENT\n-  67,000          1000                2   

              Deposit    Bathrooms            Facing  \
33957      No Deposit  3 bathrooms      South facing   
45489      No Deposit  2 bathrooms               NaN   
55208      No Deposit  3 bathrooms               NaN   
10108      No Deposit  3 bathrooms  NorthEast facing   
66484      No Deposit  4 bathrooms               NaN   
...               ...          ...               ...   
67496  85,000 deposit  3 bathrooms               NaN   
26662      No Deposit  2 bathrooms               NaN   
30800      No Deposit  1 bathrooms       East facing   
29682      No Deposit  2 bathrooms               NaN   
46055      No Deposit  2 bathrooms               NaN   

                          Address          Apartment Name          Status  
33957           Model Town, Delhi                     NaN       Furnished  
45489       Mansa Ram Park, Delhi                     NaN  Semi-Furnished  
55208     Nanakramguda, Hyderabad      Pacifica Hillcrest     Unfurnished  
10108              Wadala, Mumbai              Bombay ICC       Furnished  
66484           New Town, Kolkata  Sanjeeva Sanjeeva Town  Semi-Furnished  
...                           ...                     ...             ...  
67496            Nikol, Ahmedabad                     NaN  Semi-Furnished  
26662  byatarayanapura, Bangalore                     NaN  Semi-Furnished  
30800  Krishnarajapura, Bangalore                     NaN       Furnished  
29682  byatarayanapura, Bangalore                     NaN  Semi-Furnished  
46055    Sarvodaya Enclave, Delhi                     NaN       Furnished  

[150 rows x 10 columns]

In [14]:
duplicated_rows.shape

(71650, 10)

In [15]:
82016-68994

13022

#### 3.3 Observations

- The type of `Rent`, `Deposit`, `bathroom`should be changed to numeric type.
- There're 6700 duplicates. These should be removed.

### 4.DETAILED ANALYSIS

#### broker_consultancy


In [16]:
data.broker_consultancy

0        Kasturi Developers\nBUILDER\n-
1        Kasturi Developers\nBUILDER\n-
2        Kasturi Developers\nBUILDER\n-
3                Seller\nVERIFIED OWNER
4                Seller\nVERIFIED OWNER
                      ...              
82011            Seller\nVERIFIED OWNER
82012            Seller\nVERIFIED OWNER
82013            Seller\nVERIFIED OWNER
82014            Seller\nVERIFIED OWNER
82015            Seller\nVERIFIED OWNER
Name: broker_consultancy, Length: 82016, dtype: object

In [17]:
data.broker_consultancy.dtypes

dtype('O')

In [18]:
data.broker_consultancy.isna().sum()

1

In [19]:
data.broker_consultancy.unique()

array(['Kasturi Developers\nBUILDER\n-', 'Seller\nVERIFIED OWNER',
       'CHERRISH ESTATE CONSULTANTS\nAGENT\n-', ...,
       'ROYALTY SOLUTION SERVICES\nAGENT\n0',
       'NARAYAN REALTORS GUWAHATI\nAGENT\n0', 'DEBOJIT\nAGENT\n0'],
      dtype=object)

-   The data type of broker_consultancy is valid
-   broker_consultancy contain one null value
-   There is no used of this feature in model_trainng
-   we are going drop this in data_cleaning operation




#### Rent

In [20]:
data.Rent

0        20,000
1        20,000
2        30,000
3        28,500
4        50,000
          ...  
82011    11,000
82012     9,000
82013    22,000
82014    11,000
82015    40,000
Name: Rent, Length: 82016, dtype: object

In [21]:
(
    data
    .Rent
    .loc[lambda ser: ser.str.contains("[^0-9,]")]
)

23       1.65 L
26        2.8 L
27        1.5 L
30        1.3 L
36        1.4 L
          ...  
81255       1 L
81275       1 L
81295       1 L
81315       1 L
81335       1 L
Name: Rent, Length: 4822, dtype: object

In [22]:
data.Rent.isna().sum()

0

In [23]:
data.Rent.dtypes

dtype('O')

In [24]:
(
    data
    .Rent
    .loc[lambda ser: ser.str.contains("[^0-9,]")]
     .str.replace("L","")
     .astype(float)
      .mul(100000)
      .astype(int)
      
    
    
)

23       165000
26       280000
27       150000
30       130000
36       140000
          ...  
81255    100000
81275    100000
81295    100000
81315    100000
81335    100000
Name: Rent, Length: 4822, dtype: int64

-   The data type of Rent  is not valid
-   In rent column data contain alphanumeric values like 1.5L,3.5L values
-   we need to convert this into the numeric value
-   This feature contain zero null value



#### Builtup_Area

In [ ]:
indices = data.index[data['area'] == 8326].tolist()
print("\nIndices of rows where 'area' is 8326:")
print(indices)


-   The data type of Rent  is  valid
-   Does not contains any null value


#### Appartment_type

In [25]:
data.appartment_type.dtypes

dtype('int64')

In [26]:
data.appartment_type.isna().sum()

0

In [59]:
data.appartment_type.value_counts()

appartment_type
2    29133
3    25329
1    20777
4     5453
5     1164
6      125
7       32
8        2
9        1
Name: count, dtype: int64

-   The data type of Rent  is  valid
-   Does not contains any null value
-   we are going to remove the apparttment type which has ocuurance  less than 150 

#### Deposit

In [27]:
data.Deposit.unique()


array(['No Deposit', '4 L deposit', '1.25 L deposit', '1 L deposit',
       '60,000 deposit', '70,000 deposit', '3 L deposit', '3.5 L deposit',
       '1.5 L deposit', '3.3 L deposit', '2 L deposit', '6 L deposit',
       '9.75 L deposit', '10 L deposit', '9 L deposit', '80,000 deposit',
       '42,000 deposit', '75,000 deposit', '5 L deposit',
       '40,000 deposit', '2.5 L deposit', '5.4 L deposit', '20 L deposit',
       '15 L deposit', '18 L deposit', '8.5 L deposit', '2.1 L deposit',
       '1.74 L deposit', '1.35 L deposit', '28,000 deposit',
       '1.67 L deposit', '1.2 L deposit', '1.16 L deposit',
       '9.9 L deposit', '3.6 L deposit', '6.2 L deposit', '2.2 L deposit',
       '4.8 L deposit', '1.6 L deposit', '2.6 L deposit', '1.7 L deposit',
       '1.3 L deposit', '81,000 deposit', '84,000 deposit',
       '69,000 deposit', '1.08 L deposit', '1.02 L deposit',
       '1.56 L deposit', '54,000 deposit', '1.05 L deposit',
       '50,000 deposit', '7,000 deposit', '30,000 de

In [28]:
(
    data
    .Deposit
    .loc[lambda ser: ser.str.contains("[^No Deposit]")]
     .count()
    
)

12653

-   The data type of Deposit   is not  valid
-   This feature have lots of value as "no_Deposit"
-   because of that we are going to drop this feature

#### Bathrooms	

In [29]:
data.Bathrooms	

0        2 bathrooms
1        2 bathrooms
2        3 bathrooms
3        2 bathrooms
4        3 bathrooms
            ...     
82011    2 bathrooms
82012    2 bathrooms
82013    4 bathrooms
82014    2 bathrooms
82015    3 bathrooms
Name: Bathrooms, Length: 82016, dtype: object

In [30]:
data.Bathrooms.isna().sum()

344

In [31]:
(
    data
    .Bathrooms
     .str.strip()
     .str.get(0)
      
     
     
    
)

0        2
1        2
2        3
3        2
4        3
        ..
82011    2
82012    2
82013    4
82014    2
82015    3
Name: Bathrooms, Length: 82016, dtype: object

-   The data type of Bathrooms is not  valid
-   it should be numeric
-   this feature contains 344 null value
-   in datacleaning operation the above problems are handle

#### Facing


-   This feature have lots of null value
-   because of that we are going to drop this feature

#### Address

In [32]:
data.Address

0                  Ulwe, Mumbai
1                  Ulwe, Mumbai
2                  Ulwe, Mumbai
3        Kandivali West, Mumbai
4               Chembur, Mumbai
                  ...          
82011       Sector 2, Faridabad
82012      Sector 91, Faridabad
82013      Sector 76, Faridabad
82014      Sector 88, Faridabad
82015      Sector 79, Faridabad
Name: Address, Length: 82016, dtype: object

In [33]:
data.Address.isna().sum()

0

In [34]:
data['Address'].str.split(',').apply(lambda x:x[1]).unique()

array([' Mumbai', '', 'Thane west', ' Pune', ' Bangalore', ' Chennai',
       ' Delhi', ' Gurgaon', ' Hyderabad', ' Kolkata', ' Ahmedabad',
       ' Ghaziabad', ' Noida', ' Indore', ' Lucknow', ' Jaipur',
       ' Bhopal', ' Vadodara', ' Surat', ' Kanpur', ' Guwahati',
       ' Faridabad'], dtype=object)

In [53]:
len(data['Address'].str.split(',').apply(lambda x:x[1]).unique())

22

In [54]:
len(data['Address'].str.split(',').apply(lambda x:x[0]).unique())

626


-   This  Address feature have loction and city 
-   we have requirement of city name
-   because of that we have to seprate out the locality and city from address
-   add new feature as city and drop the feature address from dataframe
-   This feature does not contains missing values 

#### Apartment Name 

-   This feature have lots of null value
-   because of that we are going to drop this feature

#### Status

In [35]:
data.Status

0           Unfurnished
1           Unfurnished
2        Semi-Furnished
3           Unfurnished
4        Semi-Furnished
              ...      
82011       Unfurnished
82012    Semi-Furnished
82013    Semi-Furnished
82014    Semi-Furnished
82015    Semi-Furnished
Name: Status, Length: 82016, dtype: object

In [36]:
data.Status.isna().sum()

0

In [37]:
data.Status.unique()

array(['Unfurnished', 'Semi-Furnished', 'Furnished'], dtype=object)

-   This Feature has valid datatype
-   Does not contains any null value
-   not required any cleaning opertion

### 5.DATA CLEANING OPERATION


In [38]:
 def convert_rent(rent):
        if pd.isna(rent):  # Check for NaN values
            return np.nan
        if "L" in rent:
            return int(float(rent.replace('L', '').replace(',', '').strip()) * 100000)
        else:
            return int(float(rent.replace(',', '').strip()))

In [170]:
def clean_data(df):
    # Drop the 'broker_consultancy' column and rename columns to lowercase
    df_cleaned = (
        df
         
        .drop(columns="broker_consultancy")  # Drop the 'broker_consultancy','Deposit' column
        .drop(columns="Deposit")
        .drop(columns="Facing")
        .drop(columns="Apartment Name")
        #.drop_duplicates()
        
        .rename(columns=str.lower)            # Rename columns to lowercase
        .assign
        ( bathrooms=lambda df_:
         (
            df_
            
         .bathrooms
         .str.strip()
          .str.get(0)   
             
         )
            
        )
   
    )

    

    

    # Apply the convert_rent function to the 'rent' column
    df_cleaned['rent'] = df_cleaned['rent'].apply(convert_rent)

    if 'bathrooms' in df_cleaned.columns:
        mode_bathrooms = df_cleaned['bathrooms'].mode()[0]  # Calculate mode
        df_cleaned['bathrooms'] = df_cleaned['bathrooms'].fillna(mode_bathrooms) 

    # Convert 'bathrooms' column to integer type
    df_cleaned['bathrooms'] = df_cleaned['bathrooms'].astype(int)

    df_cleaned['city']=df_cleaned['address'].str.split(',').apply(lambda x:x[1])
    df_cleaned = df_cleaned.drop(columns="address")
    
    # Rename columns by removing or replacing substrings
    df_cleaned.columns = df_cleaned.columns.str.replace('area', '')
    df_cleaned.columns = df_cleaned.columns.str.replace('builtup', 'size')
    df_cleaned.columns = df_cleaned.columns.str.replace('appartment_type', 'type')
    df_cleaned.columns = df_cleaned.columns.str.replace('status', 'furnished_type')

    if 'type' in df_cleaned.columns:
        value_counts = df_cleaned['type'].value_counts()
        types_to_keep = value_counts[value_counts >= 150].index
        df_cleaned = df_cleaned[df_cleaned['type'].isin(types_to_keep)]

    df_cleaned['city'] =  df_cleaned['city'].str.strip()
    df_cleaned['city'] = df_cleaned['city'].replace('Thane west', 'Mumbai')
    df_cleaned['city'] = df_cleaned['city'].replace('','Mumbai')
    
    

    
     
    

    


    


    
    




    

    return df_cleaned
    


In [171]:
cleandata=clean_data(data)


In [172]:
cleandata.info()

<class 'pandas.core.frame.DataFrame'>
Index: 81856 entries, 0 to 82015
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   rent            81856 non-null  int64 
 1   size            81856 non-null  int64 
 2   type            81856 non-null  int64 
 3   bathrooms       81856 non-null  int64 
 4   furnished_type  81856 non-null  object
 5   city            81856 non-null  object
dtypes: int64(4), object(2)
memory usage: 4.4+ MB


In [173]:
cleandata.describe()

rent    size      type  bathrooms
count   81856.00 81856.00 81856.00   81856.00
mean    39035.96  1240.55     2.23       2.16
std     49955.71   750.24     0.95       0.97
min      1600.00   150.00     1.00       1.00
25%     15000.00   750.00     1.00       1.00
50%     25000.00  1100.00     2.00       2.00
75%     45000.00  1500.00     3.00       3.00
max   1800000.00  8326.00     5.00       7.00

In [174]:
cleandata.sample(150)

rent  size   type  bathrooms  furnished_type       city
72395   4500    600     1          1     Unfurnished  Ghaziabad
39200  20000   1300     3          2       Furnished      Delhi
50929  13000   1000     1          1  Semi-Furnished    Gurgaon
27227  15000   1100     2          2  Semi-Furnished  Bangalore
15107  30000    508     2          2     Unfurnished     Mumbai
...      ...    ...   ...        ...             ...        ...
55837  16000   1100     2          2  Semi-Furnished  Hyderabad
23383  25000    850     2          2  Semi-Furnished  Bangalore
38265  32500   1100     3          2  Semi-Furnished      Delhi
14377  12000   1222     3          3  Semi-Furnished     Mumbai
62669  21000   1000     2          1  Semi-Furnished  Hyderabad

[150 rows x 6 columns]

In [175]:
len(cleandata.city.unique())

20

In [176]:
cleandata.city.value_counts()

city
Mumbai       16085
Delhi        14900
Hyderabad     8680
Bangalore     8246
Pune          7220
Gurgaon       6060
Ahmedabad     4760
Kolkata       3960
Noida         2820
Chennai       2033
Ghaziabad     1320
Indore        1320
Lucknow       1040
Faridabad      680
Guwahati       532
Surat          520
Jaipur         480
Vadodara       420
Kanpur         420
Bhopal         360
Name: count, dtype: int64

In [164]:
cleandata.type.value_counts()

type
2    29133
3    25329
1    20777
4     5453
5     1164
Name: count, dtype: int64

### 6.OUTLIERS DETECTIONS

In [182]:

def find_outliers(df, column_name):
    """
    Identifies outliers in the specified column of the DataFrame using the IQR method.
    """
    # Check if the column exists in the DataFrame
    if column_name not in df.columns:
        raise KeyError(f"Column '{column_name}' does not exist in the DataFrame.")
    
    # Calculate Q1 (25th percentile) and Q3 (75th percentile)
    Q1 = df[column_name].quantile(0.25)
    Q3 = df[column_name].quantile(0.75)
    # Calculate IQR
    IQR = Q3 - Q1
    # Determine outlier bounds
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    # Return boolean mask for outliers
    return (df[column_name] < lower_bound) | (df[column_name] > upper_bound)


### Rent

In [183]:

# Find outliers for 'rent'
rent_outliers = find_outliers(cleandata, 'rent')
print("Outliers for 'rent':")
print(rent_outliers)

Outliers for 'rent':
0        False
1        False
2        False
3        False
4        False
         ...  
82011    False
82012    False
82013    False
82014    False
82015    False
Name: rent, Length: 81856, dtype: bool


In [184]:
def drop_outliers(df, column_name):
    """
    Drops outliers from the specified column of the DataFrame and returns the cleaned DataFrame.
    """
    # Find outliers
    outlier_mask = find_outliers(df, column_name)
    
    # Ensure that outlier_mask is a boolean Series
    if not pd.api.types.is_bool_dtype(outlier_mask):
        raise TypeError("The outlier mask must be a boolean Series.")
    
    # Drop rows where the outlier mask is True
    df_cleaned = df[~outlier_mask]
    return df_cleaned

In [201]:
finaldata=drop_outliers(cleandata,'rent')

In [202]:
finaldata

rent  size   type  bathrooms  furnished_type       city
0      20000   1132     2          2     Unfurnished     Mumbai
1      20000   1200     2          2     Unfurnished     Mumbai
2      30000   1700     3          3  Semi-Furnished     Mumbai
3      28500    750     2          2     Unfurnished     Mumbai
4      50000   1400     3          3  Semi-Furnished     Mumbai
...      ...    ...   ...        ...             ...        ...
82011  11000   1100     2          2     Unfurnished  Faridabad
82012   9000    900     2          2  Semi-Furnished  Faridabad
82013  22000   2727     4          4  Semi-Furnished  Faridabad
82014  11000    600     2          2  Semi-Furnished  Faridabad
82015  40000   1900     3          3  Semi-Furnished  Faridabad

[76593 rows x 6 columns]

In [203]:
finaldata.describe()

rent    size      type  bathrooms
count 76593.00 76593.00 76593.00   76593.00
mean  29288.50  1147.61     2.15       2.07
std   19868.33   622.47     0.91       0.91
min    1600.00   150.00     1.00       1.00
25%   14500.00   700.00     1.00       1.00
50%   24000.00  1020.00     2.00       2.00
75%   40000.00  1350.00     3.00       2.00
max   90000.00  4200.00     5.00       6.00

### 7.Preprocessing Operations

In [204]:
finaldata

rent  size   type  bathrooms  furnished_type       city
0      20000   1132     2          2     Unfurnished     Mumbai
1      20000   1200     2          2     Unfurnished     Mumbai
2      30000   1700     3          3  Semi-Furnished     Mumbai
3      28500    750     2          2     Unfurnished     Mumbai
4      50000   1400     3          3  Semi-Furnished     Mumbai
...      ...    ...   ...        ...             ...        ...
82011  11000   1100     2          2     Unfurnished  Faridabad
82012   9000    900     2          2  Semi-Furnished  Faridabad
82013  22000   2727     4          4  Semi-Furnished  Faridabad
82014  11000    600     2          2  Semi-Furnished  Faridabad
82015  40000   1900     3          3  Semi-Furnished  Faridabad

[76593 rows x 6 columns]

-   This Dataset have city and furnished_type as categorical variable
-   we have to transform into the numeric value using Label_Encoder

In [205]:
from sklearn.preprocessing import LabelEncoder
encoder=LabelEncoder()

In [207]:
finaldata['city_encoded'] = encoder.fit_transform(finaldata['city'])
# Retrieve and print the mapping of cities to encoded values
city_mapping = dict(zip(encoder.classes_, encoder.transform(encoder.classes_)))
print("\nCity to Encoded Value Mapping:")
print(city_mapping)


City to Encoded Value Mapping:
{'Ahmedabad': 0, 'Bangalore': 1, 'Bhopal': 2, 'Chennai': 3, 'Delhi': 4, 'Faridabad': 5, 'Ghaziabad': 6, 'Gurgaon': 7, 'Guwahati': 8, 'Hyderabad': 9, 'Indore': 10, 'Jaipur': 11, 'Kanpur': 12, 'Kolkata': 13, 'Lucknow': 14, 'Mumbai': 15, 'Noida': 16, 'Pune': 17, 'Surat': 18, 'Vadodara': 19}


/tmp/ipykernel_4393/431990322.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  finaldata['city_encoded'] = encoder.fit_transform(finaldata['city'])


In [208]:
finaldata

rent  size   type  bathrooms  furnished_type       city  city_encoded
0      20000   1132     2          2     Unfurnished     Mumbai            15
1      20000   1200     2          2     Unfurnished     Mumbai            15
2      30000   1700     3          3  Semi-Furnished     Mumbai            15
3      28500    750     2          2     Unfurnished     Mumbai            15
4      50000   1400     3          3  Semi-Furnished     Mumbai            15
...      ...    ...   ...        ...             ...        ...           ...
82011  11000   1100     2          2     Unfurnished  Faridabad             5
82012   9000    900     2          2  Semi-Furnished  Faridabad             5
82013  22000   2727     4          4  Semi-Furnished  Faridabad             5
82014  11000    600     2          2  Semi-Furnished  Faridabad             5
82015  40000   1900     3          3  Semi-Furnished  Faridabad             5

[76593 rows x 7 columns]

In [209]:
finaldata['furnished_type'] = encoder.fit_transform(finaldata['furnished_type'])

/tmp/ipykernel_4393/127785236.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  finaldata['furnished_type'] = encoder.fit_transform(finaldata['furnished_type'])


In [210]:
finaldata

rent  size   type  bathrooms  furnished_type       city  city_encoded
0      20000   1132     2          2               2     Mumbai            15
1      20000   1200     2          2               2     Mumbai            15
2      30000   1700     3          3               1     Mumbai            15
3      28500    750     2          2               2     Mumbai            15
4      50000   1400     3          3               1     Mumbai            15
...      ...    ...   ...        ...             ...        ...           ...
82011  11000   1100     2          2               2  Faridabad             5
82012   9000    900     2          2               1  Faridabad             5
82013  22000   2727     4          4               1  Faridabad             5
82014  11000    600     2          2               1  Faridabad             5
82015  40000   1900     3          3               1  Faridabad             5

[76593 rows x 7 columns]

In [212]:
finaldata = finaldata.drop(columns='city')

In [213]:
finaldata

rent  size   type  bathrooms  furnished_type  city_encoded
0      20000   1132     2          2               2            15
1      20000   1200     2          2               2            15
2      30000   1700     3          3               1            15
3      28500    750     2          2               2            15
4      50000   1400     3          3               1            15
...      ...    ...   ...        ...             ...           ...
82011  11000   1100     2          2               2             5
82012   9000    900     2          2               1             5
82013  22000   2727     4          4               1             5
82014  11000    600     2          2               1             5
82015  40000   1900     3          3               1             5

[76593 rows x 6 columns]

In [214]:
finaldata.corr()

rent  size   type  bathrooms  furnished_type  city_encoded
rent            1.00   0.58  0.49       0.57           -0.12          0.15
size            0.58   1.00  0.75       0.81           -0.03         -0.02
type            0.49   0.75  1.00       0.83            0.00         -0.03
bathrooms       0.57   0.81  0.83       1.00            0.07          0.02
furnished_type -0.12  -0.03  0.00       0.07            1.00          0.19
city_encoded    0.15  -0.02 -0.03       0.02            0.19          1.00

In [218]:
model_data= finaldata.sample(frac=1, random_state=42).reset_index(drop=True)


In [219]:
model_data

rent  size   type  bathrooms  furnished_type  city_encoded
0      35000   1200     3          2               0             4
1      36000    785     2          2               1            15
2      50000   2050     3          3               1             9
3      24500    550     1          1               0             9
4      21500    600     1          1               0            15
...      ...    ...   ...        ...             ...           ...
76588  35000   1200     3          2               0             4
76589  45000    600     2          1               1            15
76590  56000   1885     3          3               1             9
76591  70000    923     2          2               1            15
76592  41000   1250     2          2               1            17

[76593 rows x 6 columns]

### 8.Split the data

In [221]:
X = model_data.drop(columns="rent")
y = model_data.rent.copy()

In [222]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [223]:
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(61274, 5) (61274,)
(15319, 5) (15319,)


### 9.Train The model

In [224]:
from sklearn.linear_model import LinearRegression

model_lr = LinearRegression()
model_lr.fit(X_train, y_train)

LinearRegression()

In [225]:
from sklearn.metrics import r2_score
r2 = r2_score(y_test, model_lr.predict(X_test))
r2

0.4110581382350328

In [227]:
# get the predicted data
y_pred = model_lr.predict(X_test)

# getting the true values
y_true = y_test

In [228]:
from sklearn.metrics import mean_absolute_error

# mean absolute error (MAE)
mae = mean_absolute_error(y_true, y_pred)
mae

11244.873757844676

In [229]:
from sklearn.metrics import mean_squared_error

# mean squared error (MSE)
mse = mean_squared_error(y_true, y_pred)
mse

231597455.08557147

In [230]:
from sklearn.metrics import r2_score

# R-squared score
r2 = r2_score(y_true, y_pred)
r2

0.4110581382350328

#### polynomial Regression 

In [231]:
from sklearn.preprocessing import PolynomialFeatures

# create a encoder object
polynomial_features = PolynomialFeatures(degree=5)

# preprocess the data to get the polynomials of X
x_polynomial = polynomial_features.fit_transform(X_train)
x_polynomial

array([[1.000000e+00, 9.500000e+02, 2.000000e+00, ..., 1.965200e+04,
        1.670420e+05, 1.419857e+06],
       [1.000000e+00, 5.000000e+02, 1.000000e+00, ..., 8.788000e+03,
        5.712200e+04, 3.712930e+05],
       [1.000000e+00, 7.500000e+02, 1.000000e+00, ..., 0.000000e+00,
        0.000000e+00, 2.430000e+02],
       ...,
       [1.000000e+00, 2.000000e+03, 1.000000e+00, ..., 1.000000e+00,
        1.000000e+00, 1.000000e+00],
       [1.000000e+00, 1.300000e+03, 2.000000e+00, ..., 7.290000e+02,
        6.561000e+03, 5.904900e+04],
       [1.000000e+00, 1.020000e+03, 3.000000e+00, ..., 0.000000e+00,
        0.000000e+00, 1.024000e+03]])

In [232]:
### create Linear Regression Model
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(x_polynomial, y_train)

LinearRegression()

In [233]:
# preprocess the data to get the polynomials of X
x_polynomial_test = polynomial_features.fit_transform(X_test)
x_polynomial_test

array([[1.000000e+00, 1.300000e+03, 3.000000e+00, ..., 4.913000e+03,
        8.352100e+04, 1.419857e+06],
       [1.000000e+00, 1.100000e+03, 3.000000e+00, ..., 8.788000e+03,
        5.712200e+04, 3.712930e+05],
       [1.000000e+00, 1.300000e+03, 2.000000e+00, ..., 7.290000e+02,
        6.561000e+03, 5.904900e+04],
       ...,
       [1.000000e+00, 1.020000e+03, 3.000000e+00, ..., 0.000000e+00,
        0.000000e+00, 1.024000e+03],
       [1.000000e+00, 2.800000e+03, 4.000000e+00, ..., 6.400000e+01,
        2.560000e+02, 1.024000e+03],
       [1.000000e+00, 6.000000e+02, 2.000000e+00, ..., 1.000000e+00,
        1.000000e+00, 1.000000e+00]])

In [234]:
from sklearn.metrics import r2_score

r2 = r2_score(y_test, model.predict(x_polynomial_test))
r2

0.3978434991045242

In [241]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import AdaBoostRegressor, RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
from xgboost import XGBRegressor
from catboost import CatBoostRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR
from sklearn.naive_bayes import GaussianNB
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline


In [243]:
# Initialize models
models = {
    'Linear Regression': LinearRegression(),
    'Decision Tree': DecisionTreeRegressor(random_state=42),
    'XGBoost': XGBRegressor(random_state=42),
    'CatBoost': CatBoostRegressor(silent=True, random_state=42),
    'AdaBoost': AdaBoostRegressor(random_state=42),
    'Random Forest': RandomForestRegressor(random_state=42),
    
}

# Train and evaluate models
results = {}

for name, model in models.items():
    # Train the model
    model.fit(X_train, y_train)
    
    # Make predictions
    y_pred = model.predict(X_test)
    
    # Evaluate the model
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    mae = mean_absolute_error(y_test, y_pred)
    
    # Store the results
    results[name] = {
        'Mean Squared Error': mse,
        'R-squared': r2,
        'Mean Absolute Error': mae
    }

# Print the evaluation results
print("Model Evaluation Report:")
for name, metrics in results.items():
    print(f"\n{name}:")
    for metric, value in metrics.items():
        print(f"{metric}: {value:.4f}")

Model Evaluation Report:

Linear Regression:
Mean Squared Error: 231597455.0856
R-squared: 0.4111
Mean Absolute Error: 11244.8738

Decision Tree:
Mean Squared Error: 59034029.1247
R-squared: 0.8499
Mean Absolute Error: 2722.4265

XGBoost:
Mean Squared Error: 54843772.6932
R-squared: 0.8605
Mean Absolute Error: 2919.0184

CatBoost:
Mean Squared Error: 54321926.4595
R-squared: 0.8619
Mean Absolute Error: 3213.4955

AdaBoost:
Mean Squared Error: 186493165.1095
R-squared: 0.5258
Mean Absolute Error: 11047.8221

Random Forest:
Mean Squared Error: 56369835.1415
R-squared: 0.8567
Mean Absolute Error: 2715.0424
